### PR TITLE
sip_proxy: SIP protocol support in envoy

### PR DESCRIPTION
**Pull request by dorisd0102**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: SIP protocol support in envoy
[Issues]: https://github.com/envoyproxy/envoy/issues/16247
Additional Description:
Risk Level: Medium
Testing: unit test & manual testing
Docs Changes: add sip proxy docs
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

@durd07 @nearbyfly @dorisd0102

### CODE REVIEW

**Comment by ggreenway** [comment]
Regarding the docs error, this should be resolved soon. See https://github.com/envoyproxy/envoy/pull/17826#issuecomment-918342441.

**Comment by dorisd0102** [comment]
@mattklein123 
This PR is a contrib for sip proxy, could you help to review? Thanks.

**Comment by mattklein123** [comment]
Please remove WIP from the title, merge main, make it pass CI, etc. and then I will take a look.

/wait

**Comment by dorisd0102** [comment]
Hi @ggreenway, 

the doc build failed in this PR.
I refer to the https://github.com/envoyproxy/envoy/pull/17826 and do some changes.
But still reports following warning, could you help to check?  Many thanks.

/tmp/tmpav26zfxu/generated/rst/api-v3/config/listener/v3/listener_components.proto.rst:108: WARNING: undefined label: extension_envoy.filters.network.sip_proxy
/tmp/tmpav26zfxu/generated/rst/intro/arch_overview/security/secpos_requires_trusted_downstream_and_upstream.rst:15: WARNING: undefined label: extension_envoy.filters.network.sip_proxy
/tmp/tmpav26zfxu/generated/rst/intro/arch_overview/security/secpos_requires_trusted_downstream_and_upstream.rst:15: WARNING: undefined label: extension_envoy.filters.network.sip_proxy

File: contrib/sip_proxy/filters/network/source/: [hunk 1]
```diff
@@ -0,0 +1,134 @@
+#include "contrib/sip_proxy/filters/network/source/config.h"
```
Comment (dio) [inline contrib/sip_proxy/filters/network/source/::1]: Sorry, seems like this file is accidentally added?

Comment (dorisd0102) [reply contrib/sip_proxy/filters/network/source/:]: @dio  Yes. I just remove this file. Thanks.

File: source/common/common/logger.h [hunk 2]
```diff
@@ -76,7 +76,8 @@ namespace Logger {
   FUNCTION(tracing)                                                                                \
   FUNCTION(upstream)                                                                               \
   FUNCTION(udp)                                                                                    \
-  FUNCTION(wasm)
+  FUNCTION(wasm)                                                                                   \
+  FUNCTION(sip)
```
Comment (mattklein123) [inline source/common/common/logger.h:80]: We don't have a good solution for this right now but I would prefer to not add a new logging category for a contrib extension. Please remove this and use something generic for now like filter, udp, etc.

File: source/extensions/filters/network/well_known_names.h [hunk 3]
```diff
@@ -59,6 +59,8 @@ class NetworkFilterNameValues {
   const std::string ZooKeeperProxy = "envoy.filters.network.zookeeper_proxy";
   // WebAssembly filter
   const std::string Wasm = "envoy.filters.network.wasm";
+  // Sip proxy filter
+  const std::string SipProxy = "envoy.filters.network.sip_proxy";
```
Comment (mattklein123) [inline source/extensions/filters/network/well_known_names.h:63]: We are trying to remove this file, so please don't add this and just inline the string where needed.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, LGTM. Please merge main and remove all of the generated_api_shadow files.

/wait

File: source/common/common/logger.h (hunk 2 above)
Comment (dorisd0102) [reply source/common/common/logger.h]: removed.

File: source/extensions/filters/network/well_known_names.h (hunk 3 above)
Comment (dorisd0102) [reply source/extensions/filters/network/well_known_names.h]: Done. Move to other file.

**Comment by dorisd0102** [comment]
@mattklein123 Many thanks for your comments. All are addressed, please help to have a further review. Thanks.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by mattklein123** [comment]
@dorisd0102 one of these tests fails for me locally: 

```
[ RUN      ] SipConnectionManagerTest.UpstreamData
TestRandomGenerator running with seed 1308786792
[2021-09-28 17:11:03.443][12][error][connection] [contrib/sip_proxy/filters/network/source/conn_manager.cc:316] sip response application error: MockResponseDecoderAppException
[2021-09-28 17:11:03.444][12][error][connection] [contrib/sip_proxy/filters/network/source/conn_manager.cc:323] [C6] sip response error: MockResponseDecoderEnvoyException
[2021-09-28 17:11:03.444][12][critical][assert] [contrib/sip_proxy/filters/network/source/conn_manager.cc:155] assert failure: metadata_ != nullptr.
[2021-09-28 17:11:03.444][12][critical][backtrace] [./source/server/backtrace.h:104] Caught Aborted, suspect faulting address 0x3e80000000c
[2021-09-28 17:11:03.444][12][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2021-09-28 17:11:03.444][12][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.20.0-dev/test/DEBUG/BoringSSL
[2021-09-28 17:11:03.476][12][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x3dc7d2c]
[2021-09-28 17:11:03.476][12][critical][backtrace] [./source/server/backtrace.h:96] #1: __restore_rt [0x7fb1b0a6b980]
[2021-09-28 17:11:03.509][12][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::Extensions::NetworkFilters::SipProxy::SipConnectionManagerTest::upstreamDataTest() [0x21969ec]
[2021-09-28 17:11:03.541][12][critical][backtrace] [./source/server/backtrace.h:96] #3: Envoy::Extensions::NetworkFilters::SipProxy::SipConnectionManagerTest_UpstreamData_Test::TestBody() [0x21900d8]
[2021-09-28 17:11:03.572][12][critical][backtrace] [./source/server/backtrace.h:96] #4: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x52cfbd4]
[2021-09-28 17:11:03.604][12][critical][backtrace] [./source/server/backtrace.h:96] #5: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x52be40b]
[2021-09-28 17:11:03.636][12][critical][backtrace] [./source/server/backtrace.h:96] #6: testing::Test::Run() [0x52a9de3]
[2021-09-28 17:11:03.667][12][critical][backtrace] [./source/server/backtrace.h:96] #7: testing::TestInfo::Run() [0x52aa7b7]
[2021-09-28 17:11:03.699][12][critical][backtrace] [./source/server/backtrace.h:96] #8: testing::TestSuite::Run() [0x52aaf6c]
[2021-09-28 17:11:03.730][12][critical][backtrace] [./source/server/backtrace.h:96] #9: testing::internal::UnitTestImpl::RunAllTests() [0x52b6f0f]
[2021-09-28 17:11:03.762][12][critical][backtrace] [./source/server/backtrace.h:96] #10: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x52d39a4]
[2021-09-28 17:11:03.794][12][critical][backtrace] [./source/server/backtrace.h:96] #11: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x52c06bb]
[2021-09-28 17:11:03.825][12][critical][backtrace] [./source/server/backtrace.h:96] #12: testing::UnitTest::Run() [0x52b6ab3]
[2021-09-28 17:11:03.857][12][critical][backtrace] [./source/server/backtrace.h:96] #13: RUN_ALL_TESTS() [0x39e8421]
[2021-09-28 17:11:03.888][12][critical][backtrace] [./source/server/backtrace.h:96] #14: Envoy::TestRunner::RunTests() [0x39e7c10]
[2021-09-28 17:11:03.919][12][critical][backtrace] [./source/server/backtrace.h:96] #15: main [0x39e5dee]
[2021-09-28 17:11:03.920][12][critical][backtrace] [./source/server/backtrace.h:96] #16: __libc_start_main [0x7fb1b0689bf7]
================================================================================
```

Any quick idea what the problem might be? Otherwise I'm going to revert.

**Comment by mattklein123** [comment]
Oh hmm I think we only do release builds in CI, so this I'm guessing is a fastbuild/debug issue. @dorisd0102 can you test and fix on fastbuild/debug? Otherwise I will revert until fixed, thanks.

**Comment by durd07** [comment]
Hi @mattklein123, I create another PR to solve this UT issue, https://github.com/envoyproxy/envoy/pull/18303.
Please help to review and merge it.
Thanks.

**Pull request merged**